### PR TITLE
Feat/atomic improvements

### PR DESCRIPTION
**Pull request by LHFVII**
## Summary by CodeRabbit

* **New Features**
  * Unified swap execution path for deposit/mint/withdraw/redeem.
  * New on-chain "get NAV" hook and swap parameters to report total assets.

* **Improvements**
  * More accurate asset↔share math with explicit rounding and initial-price handling.
  * Stricter hook validation and clearer errors for invalid vault PDAs and active hook states.
  * Better fee, cap and edge-case handling across swap flows.

* **Tests**
  * Integration tests now fail fast on program registration errors.

### CODE REVIEW

File: programs/vault/src/extensions/hook_program.rs
Comment (Taylor123) [inline programs/vault/src/extensions/hook_program.rs:1]: we should name this file based on the extension it contains. So we could name it deposit_withdraw_hooks.rs or something that indicates it's for the hooks. ExtraAccountMetas is a bit of an abstract concept that won't be recognizable to all.

File: programs/vault/src/state/vault_common.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy)]
+pub enum SwapKind {
+    Deposit,
+    Mint,
+    Withdraw,
+    Redeem,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy)]
+pub struct SwapParams {
```
Comment (Taylor123) [inline programs/vault/src/state/vault_common.rs]: While these can be seen like swaps, i believe most would prefer the nomenclature to not include "swap". Perhaps we should name this "VaultAction" or something similar

File: programs/vault/src/instructions/vault_common.rs [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+        mut,
+        token::authority = user,
+        token::mint = share_mint,
+        token::token_program = share_token_program,
+    )]
+    pub user_shares_account: InterfaceAccount<'info, TokenAccount>,
+
+    pub extra_metas: Option<AccountInfo<'info>>,
+    pub protocol: Option<AccountInfo<'info>>,
+    pub hook_program: Option<AccountInfo<'info>>,
```
Comment (Taylor123) [inline programs/vault/src/instructions/vault_common.rs]: nit: let's move optional accounts to the end. We should also put a comment that this can be further optimized by moving to `remaining_accounts`, which would allow the saving of some (1 byte per optional acct) instruction data

File: programs/vault/src/instructions/vault_common.rs [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+    ctx: Context<'_, '_, '_, 'info, VaultCommon<'info>>,
+    kind: SwapKind,
+    args: SwapParams,
+) -> Result<()> {
+    ctx.accounts.vault.assert_unpaused_and_initialized()?;
+
+    match kind {
+        // User specifies exact asset amount in, receives shares.
+        // threshold_amount = min shares out (slippage).
+        SwapKind::Deposit => {
```
Comment (Taylor123) [inline programs/vault/src/instructions/vault_common.rs]: there is still a good bit of code duplication across these types. At the very least deposit/mint and withdraw/redeem should be able to share the token transfer and Vault state updates. Happy to chat through more thoughts on what is sharable

File: programs/vault/src/instructions/vault_common.rs [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+        require!(
+            return_data.len() >= 8,
+            VaultProgramError::InvalidReturnedData
+        );
+
+        let nav = u64::from_le_bytes(
+            return_data[0..8]
+                .try_into()
+                .map_err(|_| VaultProgramError::InvalidReturnedData)?,
+        );
```
Comment (Taylor123) [inline programs/vault/src/instructions/vault_common.rs:460]: should withdraw hook return value follow that of deposit hook? NAV is returned here, but the logic in the handler below makes it seem as if it is the shares to be burned.

File: programs/vault/src/instructions/vault_common.rs [hunk 5]
```diff
@@ -0,0 +1,756 @@
...
+        }
+    }
+
+    /// Returns the Token-2022 transfer fee for the given amount, or 0 for standard SPL tokens.
+    pub fn get_transfer_fees(&mut self, amount: u64) -> Result<u64> {
+        if self.asset_mint.to_account_info().owner == &spl_token::id() {
+            return Ok(0);
+        }
+        let binding = self.asset_mint.to_account_info();
+        let mint_data = binding.data.borrow();
```
Comment (Taylor123) [inline programs/vault/src/instructions/vault_common.rs]: should this use `try_borrow`?
